Creating api for gifting Hall Of Heroes Item Reference

Hus274 opened the pull request:
Fixes #8100 
### Changes

Screenshots of the new item reference:
This pull request creates a new api that is responsible for generating a formatted gifting reference for use when rewarding users with items through the hall of heroes.
#### Top

![screen shot 2016-10-10 at 6 32 11 pm](https://cloud.githubusercontent.com/assets/17734753/19256035/2d309f3c-8f18-11e6-8977-2898d9158ade.png)
#### Pets possible values

![screen shot 2016-10-10 at 6 32 23 pm](https://cloud.githubusercontent.com/assets/17734753/19256038/36773a2e-8f18-11e6-8419-b5033117d85f.png)
#### Equipped gear

![screen shot 2016-10-10 at 6 32 34 pm](https://cloud.githubusercontent.com/assets/17734753/19256047/409f780e-8f18-11e6-933b-cd64b8b002db.png)

Hus274 committed:
Creating specific api to vend item reference for gifting items to a user through the hall of heroes. closes #8100

crookedneighbor commented on website/server/controllers/api-v3/content.js:
```diff
@@ -134,4 +143,49 @@ api.getContent = {
 };
 
+/**
+ * @api {get} /api/v3/content/hallOfHeroes Get hall of heroes item path gifting reference.
```
The URL should probably be kebab case instead of camel case. hall-of-heroes

Hus274 replied:
I've changed this to not have it's own api route so nothing to do here now.

crookedneighbor commented:
The meta group might be a better home for this.

crookedneighbor commented on website/server/controllers/api-v3/content.js:
```diff
@@ -134,4 +143,49 @@ api.getContent = {
+  method: 'GET',
+  url: '/content/hallOfHeroes',
+  async handler (req, res) {
+    let content = _.cloneDeep(common.content);
```
There's nothing dynamic in this route, so you can construct this object outside of the function once and just return that instead of re-running the same operation on every request.

crookedneighbor replied:
Also, what's the reason for cloning it? Looks like everywhere you use it, you call `_.keys` which returns a new Array. There's no danger of mutating the content file.

Hus274 replied:
that makes sense. Good call.

crookedneighbor commented on website/server/controllers/api-v3/content.js:
```diff
@@ -134,4 +143,49 @@ api.getContent = {
+      });
+    });
+
+    let gear = _getFormattedItemReference('items.gear.owned', _.union(_.keys(content.gear.flat), _.flatten(mystery)), 'Boolean');
```
Why is mystery included in here?

Hus274 replied:
I included mystery here because right now it's indexed by path and the mystery gear and all other gear have the same path.

crookedneighbor commented on website/views/options/social/hall.jade:
```diff
@@ -50,7 +50,7 @@ script(type='text/ng-template', id='partials/options.social.hall.heroes.html')
           .form-group.well
             input.form-control(type='text',placeholder='Path (eg, items.pets.BearCub-Base)',ng-model='hero.itemPath')
-            small.muted Enter the <strong>item path</strong>. E.g., <code>items.pets.BearCub-Zombie</code> or <code>items.gear.owned.head_special_0</code> or <code>items.gear.equipped.head</code>. <a href='/api/v3/content' target='_blank'>See all paths here.</a> When in doubt, ask Tyler.
+            small.muted Enter the <strong>item path</strong>. E.g., <code>items.pets.BearCub-Zombie</code> or <code>items.gear.owned.head_special_0</code> or <code>items.gear.equipped.head</code>. <a href='/api/v3/content/hallOfHeroes' target='_blank'>See all paths here.</a> When in doubt, ask Tyler.
```
Let's remove the ask Tyler part.

Hus274 replied:
Done.

crookedneighbor commented:
I have mixed feelings about this. I'm not sure it _needs_ to be an API route at all, since it's something only Admins should be using. It might make more sense to render this in the hall on the frontend (hidden by default).

paglias commented:
I agree with @crookedneighbor about non needing a route for this

Hus274 commented:
Yeah, I completely agree with this. I didn't like that an api route was giving a formatted report that didn't support localization at all but wasn't quite sure what else to do. Thanks for the advice, I'll make the update so it's not using an api route.

paglias commented:
I wouldn't worry too much about localization, though we've been translating admin-only strings so far it's true that all the admins speak English and it's the language we use to speak between ourselves.

Hus274 committed:
removed 'When in doubt, ask Tyler' from the hall of heroes line

Hus274 committed:
Moving hall of heroes item path call from server side to client side

Hus274 commented:
Addressed the comments left here. Here's some screenshots of the new behavior.

### Base Screen

![screen shot 2016-10-12 at 7 14 17 pm](https://cloud.githubusercontent.com/assets/17734753/19334408/8efc6658-90b0-11e6-992f-b183335d0885.png)

### Item Paths Expanded

![screen shot 2016-10-12 at 7 14 28 pm](https://cloud.githubusercontent.com/assets/17734753/19334410/944297f4-90b0-11e6-96f8-0aad88a14ff4.png)

### Current Items Expanded

![screen shot 2016-10-12 at 7 14 37 pm](https://cloud.githubusercontent.com/assets/17734753/19334413/9d122660-90b0-11e6-9f71-161578813b8e.png)

crookedneighbor commented on website/client-old/js/controllers/hallCtrl.js:
```diff
@@ -1,16 +1,56 @@
+    }
+
+    function getAllItemPaths () {
+      let content = Content;
```
This is client side, so this will fail in old browsers.

crookedneighbor replied:
Using `let` I mean.

Hus274 replied:
updated

Hus274 committed:
updating let to var

crookedneighbor commented on website/client-old/js/controllers/hallCtrl.js:
```diff
@@ -1,16 +1,56 @@
+    }
+
+    function getAllItemPaths () {
+      var content = Content;
```
What's the reason for assigning this variable? Can't you just use `Content` everywhere?

Hus274 replied:
yep, updated.

crookedneighbor commented on website/client-old/js/controllers/hallCtrl.js:
```diff
@@ -1,16 +1,56 @@
+    function _getFormattedItemReference (pathPrefix, keys, values) {
+      var finishedString = '\n'.concat('path: ', pathPrefix, ', ', 'value: {', values, '}\n');
+
+      _.each(keys, (key) => {
```
Arrow functions will break in old browsers.

Hus274 replied:
got it

crookedneighbor commented on website/views/options/social/hall.jade:
```diff
@@ -50,12 +50,15 @@ script(type='text/ng-template', id='partials/options.social.hall.heroes.html')
+            small.muted Enter the <strong>item value</strong>. E.g., <code>5</code> or <code>false</code> or <code>head_warrior_3</code>. All values are listed in the All Item Paths section below.
+            accordion
+              accordion-group(heading='All Item Paths')
+                pre{{allItemPaths}}
```
Space after `pre` and add the `::` marker for one-time-binding

Hus274 replied:
done.

crookedneighbor commented on website/client-old/js/controllers/hallCtrl.js:
```diff
@@ -1,16 +1,56 @@
+      var pets = _getFormattedItemReference('items.pets', _.keys(content.petInfo), '-1: Owns Mount, 0: Not Owned, 1-49: Progress to mount');
+      var special = _getFormattedItemReference('items.special', _.keys(content.special), 'Numeric Quantity');
+
+      var mystery = _.flatten(_.keys(content.mystery).map((mysterySetKey) => {
```
You don't need to add the mystery sets here. The gear items are already included in the gear content object.

Hus274 replied:
Ahh gotcha, I didn't realize that was already included in the gear.

Hus274 committed:
Addressing PR comments

Hus274 reviewed (commented):
Update PR based on Comments

Hus274 commented:
Updated based on PR's comments.

TheHollidayInn commented:
@SabreCat This PR has the right idea and is a good step towards the browser I am talking about.

SabreCat commented:
@TheHollidayInn: Ah yes! It briefly came to mind when you mentioned that, but I was a doofus and didn't bring it up at the time. Definitely a worthy stopgap until we can refactor the content stuff.

paglias commented on website/client-old/js/controllers/hallCtrl.js:
```diff
@@ -1,16 +1,45 @@
+    }
+
+    function getAllItemPaths () {
+      var quests = _getFormattedItemReference('items.quests', _.keys(Content.quests), 'Numeric Quantity');
```
Heads up, we're migrating to lodash v4 (https://github.com/HabitRPG/habitica/pull/8495) which includes many breaking changes (https://github.com/lodash/lodash/wiki/Changelog#v400). Before merging this PR should be checked and updated to make sure all lodash methods are compatible with v4.

Alys commented:
I've tested in the Hall in a local install and it looks good. I scrolled through it all and didn't find anything wrong at a glance. I checked a few items at random and they were fine.

I haven't reviewed the code.

SabreCat commented:
Merged! This will go to production after a stretch of staging test. Noted toward Tier 7, thank you!
